### PR TITLE
sim_remote.cpp cannot rely on memalign(); roll our own for now

### PR DESCRIPTION
**Pull request by steven-johnson**
(For future reference: it appears that the memalign() available in the simulator runtime can return multiple overlapping allocations, which causes havoc and crashiness. Hopefully the simulator will be fixed, but in the meantime, we are using this workaround.)

### CODE REVIEW

**Comment by ronlieb** [comment]
LGTM

File: src/runtime/hexagon_remote/sim_remote.cpp [hunk 1]
```diff
@@ -99,6 +99,25 @@ __attribute__((weak)) extern int munmap;
...
+    // We want to store the original pointer prior to the pointer we return.
+    void *ptr = (void *)(((size_t)orig + alignment + sizeof(void*) - 1) & ~(alignment - 1));
+    ((void **)ptr)[-1] = orig;
+    return ptr;
+}
+
+static void aligned_free(void *ptr) {
+    if (ptr) {
+        free(((void**)ptr)[-1]);
+    };
```
Comment (abadams) [inline src/runtime/hexagon_remote/sim_remote.cpp]: trailing semicolon here

Comment (steven-johnson) [inline src/runtime/hexagon_remote/sim_remote.cpp]: done

**Comment by steven-johnson** [comment]
I just adapted this code directly from posix_allocator.cpp, but looking at it a second time, are we guaranteed to have the extra space we need for the stashed pointer? If we alloc(1) here then we might only get 1+128=129 bytes allocated, meaning we are writing before the start of the allocated space.

**Comment by pranavb-ca** [comment]
Thanks @steven-johnson.
@ronlieb @dpalermo - This could well be the reason we were getting all those failures when we tried to support ".parallel" on the simulator.

**Pull request merged**